### PR TITLE
Added new `Post-training an LLM using GRPO with TRL` recipe 🧑‍🍳️

### LINKED ISSUE #277: New `Post-training an LLM using GRPO with TRL` recipe 🧑‍🍳️
I propose adding a new recipe for post-training an LLM using GRPO with TRL, leveraging the new [GRPO Trainer available in TRL](https://huggingface.co/docs/trl/main/en/grpo_trainer).

### PR DESCRIPTION
**Pull request by sergiopaniego**
# What does this PR do?

Draft! Still in progress...

Fixes #277

## Who can review?

@merveenoyan and @stevhliu

### CODE REVIEW

**Review by qgallouedec — COMMENTED** [review]
Nice! A few remarks:

- You can now use `trl==0.14` (latest, released today) to use GRPO. Plus, datasets, accelerate and transformers are trl deps:

```diff
- !pip install  -U -q trl peft
+ 
```

- You don't need to pass the tokenizer. GRPO will load it for you
- You write:

> In the case of the DeepSeek-R1 training, they use an accuracy-based reward model to evaluate whether the response is correct, along with a format-based reward that ensures the model places its reasoning process between <think> </think> tags. You can find more details [here](https://github.com/huggingface/open-r1/blob/main/src/open_r1/grpo.py).

Why don't you use a similar reward function as well? If you remove `remove_unsused_columns=True`, you'll get access to the "solution" column of the dataset in the reward function.

File: notebooks/en/fine_tuning_llm_grpo_trl.ipynb [hunk 1]
```diff
@@ -0,0 +1,1304 @@
+{
```
Comment (burtenshaw) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p><a href="https://openai.com/index/learning-to-reason-with-llms/" target="_blank"><strong>OpenAI o1-o3 models</strong></a> &gt;&gt; <a href="https://openai.com/index/learning-to-reason-with-llms/" target="_blank"><strong>OpenAI o1 and o3 models</strong></a></p><p>exclusively employs&nbsp;<strong>pure RL &gt;&gt; </strong>exclusively employs<strong> RL |</strong> employs&nbsp;<strong>pure RL</strong></p><p>to handle more complex and nuanced tasks &gt;&gt; to handle complex and nuanced tasks</p><p> </p><p>Maybe link to the diagram in the TRL [docs](<a href="https://huggingface.co/docs/trl/main/en/grpo_trainer" target="_blank">https://huggingface.co/docs/trl/main/en/grpo_trainer</a>) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (burtenshaw) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    # Tested with transformers==4.48.1, trl==0.14.0.dev0, datasets==3.2.0, peft==0.14.0, accelerate==1.3.0
</pre><p>I think trl was just released so you can drop the dev release.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (burtenshaw) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    !pip install git+https://github.com/huggingface/trl.git@main
</pre><p>As above, I think that means we can skip installing from main.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (burtenshaw) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    print(train_dataset[0])
</pre><p>If you wanted, you could render this math with Ipython</p><p> </p>

<pre class="ql-syntax" spellcheck="false">from IPython.display import display, Math
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (burtenshaw) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p>I think this should say what the 'baseline model' is in relation to the figure above.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (burtenshaw) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p>I would move this paragraph up and then introduce this notebooks implementation. By saying something like "We will simplify this slight..."</p><p> </p><p>In the case of the DeepSeek-R1 training, they use an &gt;&gt; For training, the DeepSeek-R1 authors used an</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

**Comment by burtenshaw** [comment]
Looks really good. I left some small readability nits.

On the title, have you thought about something that relates a bit more to the task. i.e. "Post training an LLM for reasoning with GRPO in TRL" .

**Comment by qgallouedec** [comment]
Can you add the notebook to https://huggingface.co/docs/trl/en/community_tutorials when it's merged?

File: notebooks/en/fine_tuning_llm_grpo_trl.ipynb (hunk 1 above)
Comment (stevhliu) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p>I would consider leaving only the most important text/things you want to highlight in bold so as to not make it too distracting. For example, I think text like Group Relative Policy Optimization can be bold but not necessary to bold Large Language Model.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (stevhliu) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p>Let's see how and example looks like &gt;&gt; Let's take a look at an example</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (stevhliu) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p>To begin, we'll load the baseline model &gt;&gt; To begin, we'll load Qwen/Qwen2-0.5B-Instruct as the baseline model. With only 0.5 billion parameters, it is lightweight and fits within the available resources.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (stevhliu) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p>I think it'd be easier to follow if we do something like:</p><p> </p><p>In this case, we will use two reward functions. The first reward function assigns higher scores to longer completions.</p><p> </p><p>&lt;code for length_reward function here&gt;</p><p> </p><p>The second reward function ensures the generation follows a specific format, using ...</p><p> </p><p>&lt;code for format_reward function here&gt;</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (stevhliu) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p>we pass a list of reward functions to the trainer that we previously defined &gt;&gt;&gt; we pass the two reward functions we previously defined to the trainer</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (stevhliu) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p>&gt;&gt;&gt; Time to train the model!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

Comment (stevhliu) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p>anser &gt;&gt;&gt; answer</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

**Review by stevhliu — COMMENTED** [review]
Great job!

**Comment by sergiopaniego** [comment]
Thanks a lot for the feedback @qgallouedec @burtenshaw @stevhliu!! Really interesting suggestions that I believe improve the overall quality a lot 😄 

I've incorporated your feedback and updated the recipe accordingly. Following @qgallouedec's suggestion, I introduced a third reward function for accuracy, though the results and conclusions remain largely similar. I've also restructured some sections and expanded the final part with observations that I believe will be relevant to readers.

Let me know your thoughts! 😊

File: notebooks/en/fine_tuning_llm_grpo_trl.ipynb [hunk 2]
```diff
@@ -0,0 +1,3950 @@
...
```
Comment (qgallouedec) [inline notebooks/en/fine_tuning_llm_grpo_trl.ipynb:1]: <p>Just to be clear, even if in the R1 paper, the completion length increases, there is no incentive, like explicit reward for this.</p><p>I think you should remove this length reward (the completion length is logged anyway if you want to monitor the completion length)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/278/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

**Comment by sergiopaniego** [comment]
Thanks for the feedback @qgallouedec! 😄 
I've removed the length reward function to better align with R1. Initially, I included slightly different reward functions to demonstrate that the trainer could handle them. However, since we've now added the two "official" reward functions, I agree that simplifying and sticking to just those makes more sense.

**Review by stevhliu — APPROVED** [review]
LGTM, thanks again! 👏
